### PR TITLE
[RegAllocFast] Replace UsedInInstr with vector

### PR DESCRIPTION
**Pull request by aengelke**
A SparseSet adds an avoidable layer of indirection and possibly looping control flow. Avoid this overhead by using a vector to store UsedInInstrs and PhysRegUses.

To avoid clearing the vector after every instruction, use a monotonically increasing counter. The two maps are now merged and the lowest bit indicates whether the use is relevant for the livethrough handling code only.

---

http://llvm-compile-time-tracker.com/compare.php?from=739a9605677dd736971b17a7888f9d18fd245904&to=86e8efec8f9e09e7ed0aaaf47d5bea3fe793b1f6&stat=instructions:u

Not really related: iterating over regunits isn't fast. I assume (but haven't verified) that this comes from data dependencies from the diff list storage. I think a more efficient way to map regs to reg units, maybe even by reordering enum values, would improve performance. (It's not really slow either, but it happens extremely often and is therefore noticable.)

### CODE REVIEW

File: llvm/lib/CodeGen/RegAllocFast.cpp [hunk 1]
```diff
@@ -253,11 +253,19 @@ class RegAllocFastImpl {
...
 
-  using RegUnitSet = SparseSet<uint16_t, identity<uint16_t>>;
-  /// Set of register units that are used in the current instruction, and so
+  /// Track register units that are used in the current instruction, and so
   /// cannot be allocated.
-  RegUnitSet UsedInInstr;
-  RegUnitSet PhysRegUses;
+  ///
+  /// In the first phase (tied defs/early clobber), we consider also physical
+  /// uses, afterwards, we don't. if the lowest bit isn't set, it's a solely
```
Comment (arsenm) [inline llvm/lib/CodeGen/RegAllocFast.cpp:260]: Capitalize if

File: llvm/lib/CodeGen/RegAllocFast.cpp [hunk 2]
```diff
@@ -253,11 +253,19 @@ class RegAllocFastImpl {
...
+  /// physical use (markPhysRegUsedInInstr), otherwise, it's a normal use. To
+  /// avoid resetting the entire vector after every instruction, we track the
+  /// instruction "generation" in the remaining 31 bits -- this meands, that if
```
Comment (arsenm) [inline llvm/lib/CodeGen/RegAllocFast.cpp:263]: Typo 'meands'

File: llvm/lib/CodeGen/RegAllocFast.cpp [hunk 3]
```diff
@@ -253,11 +253,19 @@ class RegAllocFastImpl {
...
+  /// UsedInInstr[Idx] < InstrGen, the register unit is unused. InstrGen is
+  /// never zero and always incremented by two.
+  uint32_t InstrGen;
+  SmallVector<unsigned, 0> UsedInInstr;
```
Comment (arsenm) [inline llvm/lib/CodeGen/RegAllocFast.cpp:267]: Surprised by 0 size?

File: llvm/lib/CodeGen/RegAllocFast.cpp [hunk 4]
```diff
@@ -1382,7 +1389,12 @@ void RegAllocFastImpl::allocateInstruction(MachineInstr &MI) {
   // - The "free def operands" step has to come last instead of first for tied
   //   operands and early-clobbers.
 
-  UsedInInstr.clear();
+  InstrGen += 2;
+  // In the event we ever get more than 2**31 instructions...
+  if (InstrGen == 0) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/RegAllocFast.cpp:1394]: Is it worth putting an unlikely here?

File: llvm/lib/CodeGen/RegAllocFast.cpp (hunk 3 above)
Comment (aengelke) [reply llvm/lib/CodeGen/RegAllocFast.cpp:270]: There are only very few targets with a reasonably small number of reg units (smallest I think is BPF with 12, but most have ~100, AMDGPU over 1k) -- adding an almost never used inline storage doesn't really make sense and the allocation is done just once per module.

File: llvm/lib/CodeGen/RegAllocFast.cpp (hunk 4 above)
Comment (aengelke) [reply llvm/lib/CodeGen/RegAllocFast.cpp]: Didn't change assembly, but added it anyway for clarity.

**Review by arsenm — APPROVED** [review]

**Pull request merged**